### PR TITLE
ENG-729: add support for anonymous puzzles and steps

### PR DESCRIPTION
**Pull request by stormcloud266**
- regen sdl for puzzles
- add component tests

### CODE REVIEW

File: api/src/services/ik/steps/steps.ts [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+  return getOptionalStep({ id, puzzleId, stepNum, context })
+}
+
+// this is skipAuth
+export const anonOptionalStep: QueryResolvers['anonOptionalStep'] = async (
+  { id, puzzleId, stepNum },
+  { context }
+) => {
+  return getOptionalStep({ id, puzzleId, stepNum, context })
+}
```
Comment (stormcloud266) [inline api/src/services/ik/steps/steps.ts:62]: Turns out I did need both so we could keep our old optionalStep query behind requireAuth

File: api/src/graphql/ik/attempts.sdl.ts [hunk 2]
```diff
@@ -3,29 +3,12 @@ export const schema = gql`
...
-  type StepProgress {
-    stepId: String!
-    solved: Boolean!
-    stepSortWeight: Int!
-  }
-
-  type StepsByPuzzleIdResponse {
-    puzzle: Puzzle!
-    stepProgress: [StepProgress!]!
-  }
```
Comment (stormcloud266) [inline api/src/graphql/ik/attempts.sdl.ts:16]: Just moving our step stuff out of these attempts files

File: api/src/services/ik/attempts/attempts.ts [hunk 3]
```diff
@@ -215,37 +210,3 @@ export const makeAttempt: MutationResolvers['makeAttempt'] = async ({
...
-) => {
-  // users don't need specific step data on the puzzle landing page
-  if (!id) return
-
-  // unauthenticated users can't see specific step data
-  if (!isAuthenticated()) {
-    throw new ForbiddenError('must sign in')
-  }
-
-  // authenticated users should be allowed to view first steps
```
Comment (stormcloud266) [inline api/src/services/ik/attempts/attempts.ts:236]: This whole function was moved to `steps` but I also deleted this check and made the non-anon query requireAuth

File: web/src/components/StepsCell/StepsCell.tsx [hunk 4]
```diff
@@ -71,66 +67,22 @@ export const Success = ({
...
-            })}
-            to={
-              isAuthenticated
-                ? routes.puzzleStep({
-                    slug,
-                    step: stepSortWeight,
-                  })
-                : null
-            }
-          />
```
Comment (stormcloud266) [inline web/src/components/StepsCell/StepsCell.tsx:125]: In the `StepsCell` this `isAuthenticated` check now just routes to `anonPuzzle` if they aren't authenticated

File: api/src/graphql/ik/steps.sdl.ts [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+
+  type StepsByPuzzleIdResponse {
+    puzzle: Puzzle!
+    stepProgress: [StepProgress!]!
+  }
+
+  type Query {
+    getStepsByPuzzleId(id: String!): StepsByPuzzleIdResponse @skipAuth
+    optionalStep(id: String, puzzleId: String, stepNum: Int): Step @requireAuth
+    anonOptionalStep(id: String, puzzleId: String, stepNum: Int): Step @skipAuth
```
Comment (stormcloud266) [inline api/src/graphql/ik/steps.sdl.ts:16]: This is the new one

File: api/src/services/ik/steps/steps.ts [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+const getOptionalStep = ({
+  id,
+  puzzleId,
+  stepNum,
+  context,
+}: {
+  id: string
+  puzzleId: string
+  stepNum: number
+  context: RedwoodGraphQLContext
```
Comment (illepic) [inline api/src/services/ik/steps/steps.ts:20]: Let's use global context instead.

File: web/src/components/PuzzleLandingLayout/PuzzleLandingLayout.tsx [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+import RewardableHeader from 'src/components/RewardableHeader/RewardableHeader'
+import Seo from 'src/components/Seo/Seo'
+import TwitterShare from 'src/components/TwitterShare/TwitterShare'
+
+import '@infinity-keys/react-lens-share-button/dist/style.css'
+
+interface PuzzleLandingLayoutProps extends PropsWithChildren {
+  rewardable:
+    | NonNullable<FindAnonRewardablePuzzleBySlug['rewardable']>
+    | NonNullable<FindRewardablePuzzleBySlug['rewardable']>
```
Comment (illepic) [inline web/src/components/PuzzleLandingLayout/PuzzleLandingLayout.tsx]: Double check that `NonNullable` is needed here.

File: web/src/components/PuzzleLandingLayout/PuzzleLandingLayout.tsx [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+  FindRewardablePuzzleBySlug,
+} from 'types/graphql'
+
+import RewardableHeader from 'src/components/RewardableHeader/RewardableHeader'
+import Seo from 'src/components/Seo/Seo'
+import TwitterShare from 'src/components/TwitterShare/TwitterShare'
+
+import '@infinity-keys/react-lens-share-button/dist/style.css'
+
+interface PuzzleLandingLayoutProps extends PropsWithChildren {
```
Comment (illepic) [inline web/src/components/PuzzleLandingLayout/PuzzleLandingLayout.tsx:16]: Let's get this into a Layout/Set

File: web/src/components/StepsLayout/StepsLayout.tsx [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+}: StepsLayoutProps) => {
+  const { slug } = useParams()
+  const { isAuthenticated } = useAuth()
+
+  return (
+    <div>
+      {step && (
+        <div>
+          {step.type === 'SIMPLE_TEXT' && (
+            <SimpleTextInput
```
Comment (illepic) [inline web/src/components/StepsLayout/StepsLayout.tsx]: Try dynamic import here.

**Review by illepic — CHANGES_REQUESTED** [review]

File: api/src/services/ik/steps/steps.ts (hunk 6 above)
Comment (stormcloud266) [reply api/src/services/ik/steps/steps.ts:20]: Global context gives me this error about it being type unknown:

Screenshot 2023-03-01 at 10 26 53 AM


whereas it comes in typed getting it from the params:

Screenshot 2023-03-01 at 10 27 43 AM

File: web/src/components/PuzzleLandingLayout/PuzzleLandingLayout.tsx (hunk 8 above)
Comment (stormcloud266) [reply web/src/components/PuzzleLandingLayout/PuzzleLandingLayout.tsx:16]: So I actually have to pass in the rewardable to this layout on each page, so I don't think I'll be able to move this into a set

**Review by illepic — APPROVED** [review]

**Pull request merged**